drivers/clock_control/clock_control_nrf2_hfxo.c

```diff
@@ -0,0 +1,196 @@
+{
+	struct dev_data_hfxo *dev_data =
+		CONTAINER_OF(mgr, struct dev_data_hfxo, mgr);
+
+	dev_data->notify = notify;
+
+	nrf_lrcconf_event_clear(NRF_LRCCONF010, NRF_LRCCONF_EVENT_HFXOSTARTED);
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_REQHFXO);
+
+	k_timer_start(&dev_data->timer, dev_data->start_up_time, K_NO_WAIT);
+}
+
+static void onoff_stop_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_STOPREQHFXO);
+
+	notify(mgr, 0);
+}
+
+static bool is_clock_spec_valid(const struct device *dev,
+				const struct nrf_clock_spec *spec)
+{
+	const struct dev_config_hfxo *dev_config = dev->config;
+	uint16_t accuracy = spec->accuracy;
+
+	if (spec->frequency > dev_config->fixed_frequency) {
+		LOG_ERR("invalid frequency");
+		return false;
+	}
+
+	if (spec->precision) {
+		LOG_ERR("invalid precision");
+		return false;
+	}
+
+	if (accuracy == NRF_CLOCK_CONTROL_ACCURACY_MAX) {
+		accuracy = dev_config->max_accuracy;
+	}
+
+	if (accuracy < dev_config->max_accuracy) {
+		LOG_ERR("invalid accuracy");
+		return false;
+	}
+
+	return true;
+}
+
+static int api_request_hfxo(const struct device *dev,
+			    const struct nrf_clock_spec *spec,
+			    struct onoff_client *cli)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+
+	if (spec && !is_clock_spec_valid(dev, spec)) {
+		return -EINVAL;
+	}
+
+	return onoff_request(&dev_data->mgr, cli);
+}
+
+static int api_release_hfxo(const struct device *dev,
+			    const struct nrf_clock_spec *spec)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+
+	if (spec && !is_clock_spec_valid(dev, spec)) {
+		return -EINVAL;
+	}
+
+	return onoff_release(&dev_data->mgr);
+}
+
+static int api_cancel_or_release_hfxo(const struct device *dev,
+				      const struct nrf_clock_spec *spec,
+				      struct onoff_client *cli)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+
+	if (spec && !is_clock_spec_valid(dev, spec)) {
+		return -EINVAL;
+	}
+
+	return onoff_cancel_or_release(&dev_data->mgr, cli);
+}
+
+static int api_get_rate_hfxo(const struct device *dev,
+			     clock_control_subsys_t sys,
+			     uint32_t *rate)
+{
+	ARG_UNUSED(sys);
+
+	const struct dev_config_hfxo *dev_config = dev->config;
+
+	*rate = dev_config->fixed_frequency;
+
+	return 0;
+}
+
+static int init_hfxo(const struct device *dev)
+{
+	struct dev_data_hfxo *dev_data = dev->data;
+	static const struct onoff_transitions transitions = {
+		.start = onoff_start_hfxo,
+		.stop = onoff_stop_hfxo
+	};
+	int rc;
+
+	rc = onoff_manager_init(&dev_data->mgr, &transitions);
+	if (rc < 0) {
+		return rc;
+	}
+
+	k_timer_init(&dev_data->timer, hfxo_start_up_timer_handler, NULL);
+
+#ifdef NRF_BICR
+	uint32_t bicr_startuptime_us = NRF_BICR->HFXO.STARTUPTIME;
+
+	if (bicr_startuptime_us != BICR_HFXO_STARTUPTIME_TIME_Unconfigured) {
+		dev_data->start_up_time = K_USEC(bicr_startuptime_us);
+	} else
+#endif
+	{
+		dev_data->start_up_time = K_USEC(1000);
```

Comment:
BICR misconfiguration should be treated as a fault instead of setting a default value. This hardcoded default could be different than similar defaults in other libraries (e.g. radio protocols).